src/compiler/checker.ts

```diff
@@ -34268,7 +34273,39 @@ namespace ts {
                 const saveCurrentNode = currentNode;
                 currentNode = node;
                 instantiationCount = 0;
+
+                const oldTypeCount = typeCount;
+                const oldSymbolCount = symbolCount;
+
                 checkSourceElementWorker(node);
+
+                if (node.kind >= SyntaxKind.FirstStatement && node.kind <= SyntaxKind.LastStatement) {

```

Comment:
This could be more aggressive about skipping work if the statements will never be reported.